cpp/src/arrow/compute/exec/plan_test.cc

```diff
@@ -344,6 +344,47 @@ void TestSourceSink(
               Finishes(ResultWith(UnorderedElementsAreArray(exp_batches.batches))));
 }
 
+void TestRecordBatchReaderSourceSink(
+    std::function<Result<std::shared_ptr<RecordBatchReader>>(const BatchesWithSchema&)>
+        to_reader) {
+  ASSERT_OK_AND_ASSIGN(auto executor, arrow::internal::ThreadPool::Make(1));
+  ExecContext exec_context(default_memory_pool(), executor.get());
+  ASSERT_OK_AND_ASSIGN(auto plan, ExecPlan::Make(exec_context));
+  AsyncGenerator<std::optional<ExecBatch>> sink_gen;
+
+  auto exp_batches = MakeBasicBatches();
+  ASSERT_OK_AND_ASSIGN(std::shared_ptr<RecordBatchReader> reader, to_reader(exp_batches));
+
+  ASSERT_OK(Declaration::Sequence(
+                {
+                    {"record_batch_reader_source",
+                     RecordBatchReaderSourceNodeOptions{exp_batches.schema, reader}},
+                    {"sink", SinkNodeOptions{&sink_gen}},
+                })
+                .AddToPlan(plan.get()));
+
+  ASSERT_THAT(StartAndCollect(plan.get(), sink_gen),
+              Finishes(ResultWith(UnorderedElementsAreArray(exp_batches.batches))));
+}
```

Comment:
You are specifying an executor but you aren't verifying that it is being used.  Also, there is no test where we do not specify an executor.

Testing that the executor gets used would be a little tricky (maybe someday we can add a `TotalTasksRun` method to the executor).  You would probably need to wrap the batch generator with something that verifies it is on the I/O executor (e.g. using OwnsThisThread or something similar).

I am probably being a bit harsh here as I suspect we test some of the other sources in a similar fashion.  If we're only going to have one simple test I would prefer a test using the default instead of one specifying a custom executor as that is the more common case.

Reply:
@westonpace I am working on a wrapper to address this. Let me evaluate options and update the PR. Also, I wouldn't say this is `harsh` but `thorough` as I see other sources are not properly tested like this. I think it would be better to test them too. I will create a ticket to cover this. Thanks for thinking in this angel.

Reply:
By the way `TotalTasksRun` sounds fun, could you elaborate a bit what you expect here, I am happy to work on that task.

Reply:
Probably external to this PR. Appreciate your thoughts.

Reply:
> Are you referring to IOThreadPool being used as the default instead of NULLPTR?

What you had before might actually have been correct (I thought you were using `executor` in the options but on closer examination it was only being used in the exec_ctx).  I was saying the most common case is `RecordBatchReaderSourceNodeOptions options(reader)` where the executor is not specified and is left to be the default.  We should have our test verify that this case works correctly.  At the very least we should make sure we can run such a plan without crashing.  Long term we should check, after we ran the plan, to ensure that there were tasks submitted to the I/O thread pool.

This kind of test should not need to create a thread pool at all.  We should only need to create a thread pool when testing a custom executor in the node.

Reply:
I updated the code, is it okay?